sklearn/impute/_base.py

```diff
@@ -465,11 +465,41 @@ def transform(self, X):
             n_missing = np.sum(mask, axis=0)
             values = np.repeat(valid_statistics, n_missing)
             coordinates = np.where(mask.transpose())[::-1]
-
             X[coordinates] = values
 
         return super()._concatenate_indicator(X, X_indicator)
 
+    def inverse_transform(self, X):
+        """Convert the data back to the original representation.
+
+        Inverts the `fit_transform` operation performed on an array.
+        This operation can only be performed after :class:`SimpleImputer` is
+        used with `add_indicator` set to `True`.
+
+        Parameters
+        ----------
+        X : array-like,
+            shape (n_samples, n_features + missing_feature_count)
+            The imputed data to be reverted to original data. It has to be
+            an augmented array of imputed data and the missing indicator mask.
+        """
+        check_is_fitted(self)
+        missing_feature_count = len(self.indicator_.features_)
+
+        # Split the augmented array into imputed array and it's missing
+        # indicator mask.
+        feature_count = X.shape[1] - missing_feature_count
+        imputed_arr = X[:, :feature_count].copy()
+        missing_mask = X[:, feature_count:].copy()
+        missing_mask = missing_mask.astype(np.bool)
+
+        # Iterate over features and replace the imputed values
+        # with original missing value types.
+        for i in range(missing_feature_count):
```

Comment:
Isn't it possible to avoid the `for` loop ?
Like using directly 
```py
            imputed_arr[:, self.indicator_.features_][missing_mask] = self.missing_values
```
?

Reply:
Thank you Tom for point this out. I tried your suggestion, but it turns out that 
`imputed_arr[:, self.indicator_.features_][missing_mask]` returns a copy of the array - so the assignment does not change `imputed_arr`. A slice followed by a mask apparently returns a new array<sup>[1]</sup>, and I did not know of that!

However, I did get rid of the `for` loop. I expanded `missing_mask` to the shape of `imputed_arr` and use that for assignment.

<sup>[1]</sup> [Explained well here](https://stackoverflow.com/questions/38768815/view-of-a-view-of-a-numpy-array-is-a-copy)

Reply:
Good job !